content/learn/05.communication/07.modbus/modbus.md

```diff
@@ -0,0 +1,52 @@
+
+## How does Modbus work?
+Each Modbus message has the same structure, consisting of four basic elements which are present in each message and the sequence and order of these elements are the same for all messages. This allows for easy parsing of the content. The conversation is always started by the controller and when a message is sent the peripheral interprets the message and responds to it. Every peripheral device has its own address which it reponds to when addressed by the controller, while all other devices ignore the message if the address doesn't match with their own.
+
+Modbus sends functions which communicate read and write instructions to the peripheral’s internal memory registers to configure, monitor and control the peripheral’s inputs and outputs. Modbus devices will typically include a register map outlining where the configuration input and output data can be written and read from. You should always refer to the peripheral’s register map of your device to gain a better understanding of its overall operation. 
+
+The Modbus data model has a simple structure described in four basic data types:
+
+| Data Type         | Description   | 
+| -----------       | -----------   |
+| Discrete Input    |               |
+| Coils Input       |               |
+| Input Registers   | (Input Data)  |
+| Holding Registers | (Output Data) |
+
+
+The Modbus memory registers of a device are organized around the four basic data reference types and this data type is further identified by the leading number used in the devices memory address, such as, 
+- 0 based register referencing a message to Read or Write discrete outputs or coils, 
+- 1 based register referencing Reading discrete inputs,
+- 3 based register referencing Reading input registers,
+- 4 based register referencing Reading or Writing to output or holding registers. 
+
+![](assets/modbusStructure.png)
+
+The function code field specifies which register data group it reads or writes to and from the peripheral. Many of the data types are named from its use in driving relays, for example, a single-bit physical output is called a coil, and a single-bit physical input is called a discrete input or a contact. 
+For example, the Read Holding Registers command has the function code with 1 byte containing 8 bits is binary 0000 0011. If the peripheral device accepts the request without error, it will return the same code in its response. However, if an error occurs, the peripheral will return 1 byte containing 8 binary bits 1000 0011 in the function code field and appends a unique code in the data field of the response message that tells the controller device what kind of error occurred, or the reason for the error.
```

Comment:
```suggestion
For example, the Read Holding Registers command has the function code with 1 byte containing 8 bits is binary `0000 0011`. If the peripheral device accepts the request without error, it will return the same code in its response. However, if an error occurs, the peripheral will return 1 byte containing 8 binary bits `1000 0011` in the function code field and appends a unique code in the data field of the response message that tells the controller device what kind of error occurred, or the reason for the error.
```